astropy/units/tests/test_units.py

```diff
@@ -1022,3 +1022,23 @@ def test_dask_arrays():
 def test_get_format_name_deprecation():
     with pytest.warns(AstropyDeprecationWarning, match=r"Use to_string\(\) instead\.$"):
         assert u.m.get_format_name("fits") == "m"
+
+
+def test_comparison_dimensionless_with_np_ma_masked():
+    # Found to be a problem indirectly in gh-17047;
+    # The path np.ma.masked.__eq__(u.dimensionless_unscaled)
+    # used to give a ZeroDivisionError.
+    comparison = u.dimensionless_unscaled == np.ma.masked
+    assert comparison is np.ma.masked
+
+
+def test_error_on_conversion_of_zero_to_unit():
+    # Found to be a problem indirectly in gh-17047; we allow conversion
+    # of numbers to units, but should not allow 0.
+    with pytest.raises(u.UnitScaleError, match="cannot create.*scale of 0"):
+        u.Unit(0)
+    with pytest.raises(u.UnitScaleError, match="cannot create.*scale of 0"):
+        u.dimensionless_unscaled.to(0)
+    # Also check some that do work.
+    assert u.dimensionless_unscaled.to(0.125) == 8
+    assert u.dimensionless_unscaled.to(8) == 0.125
```

Comment:
Do we need `allclose` or this is always going to be exact across all the archs?

Reply:
I picked a power of 2 on purpose...